types/ember/test/private/observable-tests.ts

```diff
@@ -0,0 +1,175 @@
+import Observable from "@ember/object/observable";
+import { UnwrapComputedPropertyGetter, UnwrapComputedPropertyGetters, UnwrapComputedPropertySetters, UnwrapComputedPropertySetter } from "ember/-private-types/object/computed";
+// tslint-disable-next-line
+import { assertType } from "../lib/assert";
+import { Ember } from "ember";
+import { ExtractPropertyNamesOfType } from "ember/-private-types/utils";
+
+class OtherThing {
+    observerOfDemo(target: DemoObservable, key: 'foo') {}
+}
+
+class DemoObservable implements Observable {
+    foo: string;
+    isFoo = true;
+    bar: [boolean, boolean];
+    baz?: number;
+    constructor() {
+        this.foo = 'hello';
+        this.bar = [false, true];
+        this.baz = 9;
+        this.addObserver('foo', this, 'fooDidChange');
+        this.addObserver('foo', this, 'fooDidChangeProtected'); // $ExpectError
+        this.addObserver('foo', this, this.fooDidChange);
+        this.addObserver('foo', this, this.fooDidChangeProtected);
+        const ot = new OtherThing();
+        this.addObserver('foo', ot, ot.observerOfDemo);
+        Ember.addObserver(this, 'foo', this, 'fooDidChange');
+        Ember.addObserver(this, 'foo', this, 'fooDidChangeProtected'); // $ExpectError
+        Ember.addObserver(this, 'foo', this, this.fooDidChange);
+        Ember.addObserver(this, 'foo', this, this.fooDidChangeProtected);
+        this.removeObserver('foo', this, 'fooDidChange');
+        this.removeObserver('foo', this, 'fooDidChangeProtected'); // $ExpectError
+        this.removeObserver('foo', this, this.fooDidChange);
+        this.removeObserver('foo', this, this.fooDidChangeProtected);
+        Ember.removeObserver(this, 'foo', this, 'fooDidChange');
+        Ember.removeObserver(this, 'foo', this, 'fooDidChangeProtected'); // $ExpectError
+        Ember.removeObserver(this, 'foo', this, this.fooDidChange);
+        const lambda = () => {
+            this.fooDidChange(this, 'foo');
+        };
+        this.addObserver('foo', lambda);
+        this.addObserver('foo', (sender, key, value, rev) => {
+            assertType<DemoObservable>(sender);
+            assertType<'foo'>(key);
+            assertType<string>(value);
+            assertType<number>(rev);
+        });
+        this.removeObserver('foo', lambda);
+        Ember.addObserver(this, 'foo', lambda);
+        Ember.removeObserver(this, 'foo', lambda);
+    }
+
+    fooDidChange(obj: this, propName: keyof this) {}
+    protected fooDidChangeProtected(obj: this, propName: keyof this) {}
```

Comment:
All of the stuff below is just an implementation of `Observable` that throws on everything, for the purposes of testing this layer before involving `Ember.Object` or anything else.